### PR TITLE
Measured version 0.0.1

### PR DESCRIPTION
**Pull request by kmcphillips**
@Shopify/shipping @maartenvg @celsodantas @boourns @jnormore 

## What is this?

This is version 0.0.1 of a library I have written to encapsulate measurements, weight and dimensions to start with, in a single object.

Currently in Shopify we juggle two different properties which really define a single weight or dimension. We do lossy math with `UnitSystem` aka `Quantified` meaning our precision is bad. And we have no atomic way of reading/writing these to columns in the DB with ActiveRecord without juggling the fields by ourselves.

All math is done in this gem with `BigDecimal`. Conversions are extensible.

## Features

I wrote a detailed README. Read it. Please:

https://github.com/Shopify/measured/blob/version-0.0.1/README.md

## Why build from scratch?

There are two unit management libs I could find of note. They are contrasted in the README.

**tl;dr** `Quantified` in ActiveShipping does only float math and pluralizes all units. Its core assumptions are inaccurate. The `ruby-units` overrides nearly all core types and depends on those internally. Plus it is huge. Neither have ActiveRecord adapters anyway.

## Next steps

Writing a `measured-rails` adapter for ActiveRecord which understands measured objects and how to store them in columns and convert them. Inspiration for this will be taken from the Money gem.

Something roughly like:

```ruby
class Box < ActiveRecord::Base
  dimension_column :length
end
```
Which will load/store data in `length_amount` and `length_unit` columns, with optionally normalizing to `length_amount_mm` or whatever.

### CODE REVIEW

File: lib/measured/conversion_table.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    conversion
+  end
+
+  def find_direct_conversion(to:, from:)
+    @units.each do |unit|
+      return unit.conversion_amount if unit.name == from && unit.conversion_unit == to
+    end
+
+    @units.each do |unit|
+      return BigDecimal(1) / unit.conversion_amount if unit.name == to && unit.conversion_unit == from
```
Comment (robfoster) [inline lib/measured/conversion_table.rb]: not super obvious what's happening here

File: lib/measured/conversion_table.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    nil
+  end
+
+  def find_tree_traversal_conversion(to:, from:)
+    traverse(from: from, to: to, unit_names: @units.map{|u| u.name }, amount: BigDecimal("1"))
+  end
+
+  def traverse(from:, to:, unit_names:, amount:)
+    unit_names = (unit_names - [from])
```
Comment (robfoster) [inline lib/measured/conversion_table.rb]: why parenthesis?

File: lib/measured/conversion_table.rb (hunk 1 above)
Comment (kmcphillips) [inline lib/measured/conversion_table.rb]: @robfoster 

Yeah, and this is the line that I think is causing those rounding bugs.

So it creates the conversion in the mapping table in reverse: Given `x = 2 * y` then `1/2 x = y`

So if we have a mapping saying that `1 m = 100 cm` but we are asking for the opposite conversion, then we return `1 cm = 1/100 m`

Comment (kmcphillips) [inline lib/measured/conversion_table.rb]: Basically:
* Try to find a direct X -> Y conversion
* If not found, try to find the inverse Y -> X and convert it with 1/
* If not found do a recursive tree traversal to evaluate all possible paths to find an indirect conversion path

File: lib/measured/conversion_table.rb (hunk 2 above)
Comment (kmcphillips) [inline lib/measured/conversion_table.rb]: No reason. Removing.

File: lib/measured/measurable.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  include Comparable
+  include Measured::Arithmetic
+
+  attr_reader :unit, :value
+
+  def initialize(value, unit)
+    raise Measured::UnitError, "Unit #{ unit } does not exits." unless self.class.conversion.unit_or_alias?(unit)
+
+    @value = value
+    @value = BigDecimal(@value) unless @value.is_a?(BigDecimal)
```
Comment (robfoster) [inline lib/measured/measurable.rb]: alternatively, @value = value.is_a?(BigDecimal) ? value : BigDecimal(@value)

File: lib/measured/conversion_table.rb (hunk 1 above)
Comment (qq99) [inline lib/measured/conversion_table.rb]: Are you worried that repeated conversions (in a `a-b-c-...z` chain) will decrease float accuracy?  I guess for our purposes it'll be pretty accurate.  I think the inverse `1/` is a bit less accurate than doing the non-inverse conversion.

Comment (kmcphillips) [inline lib/measured/conversion_table.rb]: Absolutely. It has accuracy problems. I'm very open to suggestions.

For the most part `A -> B -> C` looks like `cm -> m -> in` where one of the steps is a rational non repeating decimal number. And the constant between `m` and `in` is the only hairy bit.

File: lib/measured/unit.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  include Comparable
+
+  def initialize(name, aliases: [], value: nil)
+    @name = name.to_s
+    @names = ([@name] + aliases.map{|n| n.to_s }).sort
+
+    @conversion_amount, @conversion_unit = parse_value(value) if value
+  end
+
+  attr_reader :name, :names, :conversion_amount, :conversion_unit
```
Comment (robfoster) [inline lib/measured/unit.rb]: order doesn't really matter, but I usually see attr_readers at the top of the class. Thoughts?

File: lib/measured/unit.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  def to_s
+    if conversion_string
+      "#{ @name } (#{ conversion_string })"
+    else
+      @name
+    end
```
Comment (robfoster) [inline lib/measured/unit.rb]: alternatively, 
```return "#{ @name } (#{ conversion_string })" if conversion_string
@name```

File: lib/measured/conversion_table.rb (hunk 1 above)
Comment (qq99) [inline lib/measured/conversion_table.rb]: Is it possible to express all conversions in rationals?  I didn't examine the code too closely but I have the feeling that you're combining rationals together for conversion

Comment (kmcphillips) [inline lib/measured/conversion_table.rb]: Could be.

Actually, I could do a check for rational here and if so swap the two values, rather than coerce to big decimal and divide one by it.

File: lib/measured/unit.rb [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+      end
+    else
+      @name <=> other
+    end
+  end
+
+  private
+
+  def conversion_string
+    "#{ conversion_amount } #{ conversion_unit }" if @conversion_amount || @conversion_unit
```
Comment (robfoster) [inline lib/measured/unit.rb]: hmm?

File: lib/measured/unit.rb [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  def parse_value(tokens)
+    tokens = tokens.split(" ") if tokens.is_a?(String)
+    raise Measured::UnitError unless tokens.size == 2
```
Comment (robfoster) [inline lib/measured/unit.rb]: Maybe: raise Measured::UnitError, "Cannot parse '#{tokens} into one value and one unit'"

File: lib/measured/unit.rb [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+    tokens[0] = BigDecimal(tokens[0]) unless tokens[0].is_a?(BigDecimal) || tokens[0].is_a?(Rational)
```
Comment (robfoster) [inline lib/measured/unit.rb]: tokens.first?

File: lib/measured/conversion_table.rb (hunk 1 above)
Comment (qq99) [inline lib/measured/conversion_table.rb]: I'm not sure which numbers would be good to test out this kind of approach (have you been testing float acc?)  Assuming ones near the gigantic scale would be best

File: lib/measured/units/dimension.rb [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+  conversion.add :ft,
+    aliases: [:foot, :feet],
+    value: "0.3048 m"
+
+  conversion.add :yd,
+    aliases: [:yard, :yards],
+    value: "0.9144 m"
+
+end
```
Comment (robfoster) [inline lib/measured/units/dimension.rb]: totally doesn't matter for the purposes of this project, but someone's going to want to extend this into kilometers and miles in the future

File: lib/measured/units/weight.rb [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    aliases: [:kilogram, :kilograms],
+    value: "1000 g"
+
+  conversion.add :lb,
+    aliases: [:lbs, :pound, :pounds],
+    value: [Rational(45359237,1e8), "kg"]
+
+  conversion.add :oz,
+    aliases: [:ounce, :ounces],
+    value: [Rational(1,16), "lb"]
```
Comment (robfoster) [inline lib/measured/units/weight.rb]: hmm, also 0.0625. Not sure 1/16 is more clear or less clear

File: lib/measured/units/dimension.rb [hunk 11]
```diff
@@ -0,0 +1,26 @@
+class Measured::Dimension < Measured::Measurable
+
+  conversion.set_base :m,
+    aliases: [:meter, :metre, :meters, :metres]
+
+  conversion.add :cm,
+    aliases: [:centimeter, :centimetre, :centimeters, :centimetres],
+    value: "0.01 m"
```
Comment (qq99) [inline lib/measured/units/dimension.rb]: Would be better to express these as `Rational(1,100)`?

File: lib/measured/units/dimension.rb [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+  conversion.add :mm,
+    aliases: [:millimeter, :millimetre, :millimeters, :millimetres],
+    value: "0.001 m"
+
+  conversion.add :in,
+    aliases: [:inch, :inches],
+    value: "0.0254 m"
```
Comment (qq99) [inline lib/measured/units/dimension.rb]: Is there a rational for this?  I don't know a lot about Imperial

File: test/arithmetic_test.rb [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+  end
+
+  test "#- should subtract a number from the value" do
+    assert_equal Magic.new(-1, :magic_missile), @two - 3
+    assert_equal Magic.new(1, :magic_missile), 2 - @three
+  end
+
+  test "#- should raise if different unit system" do
+    assert_raises TypeError do
+      OtherFakeSystem.new(1, :other_fake_base) - @two
```
Comment (robfoster) [inline test/arithmetic_test.rb]: hmm. so we can't subtract 1g from a pound? Thoughts?

File: test/arithmetic_test.rb [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+    assert_raises TypeError do
+      @two - "thing"
+    end
+
+    assert_raises NoMethodError do
+      "thing" - @two
+    end
+  end
+
+  test "#* should multiply together same units" do
```
Comment (robfoster) [inline test/arithmetic_test.rb]: what is 2g * 2g? 4g or 4 cubic g?

when we calculate dimensions we do l*w*h, and assume a cubic value.

Comment (qq99) [inline test/arithmetic_test.rb]: I think for boxes, you'd need something more like `Box.new(l: 2, w: 5, h: 10).volume`, I don't think this should multiplying grams should result in cubic

File: test/arithmetic_test.rb [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+    assert_raises TypeError do
+      @two * "thing"
+    end
+
+    assert_raises TypeError do
+      "thing" * @two
+    end
+  end
+
+  test "#/ should multiply together same units" do
```
Comment (robfoster) [inline test/arithmetic_test.rb]: s/multiply/divide/s

File: test/arithmetic_test.rb (hunk 14 above)
Comment (qq99) [inline test/arithmetic_test.rb]: I think you have an interesting point, though, when does multiplying `unit * unit` make sense?  In grams, IDK that `g^2`, `g^3` really makes sense, but volume in `cm^3` does because the 3 distances imply non-scalar values

File: test/arithmetic_test.rb [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+    assert_equal Magic.new("0.5", :magic_missile), @two / @four
```
Comment (robfoster) [inline test/arithmetic_test.rb]: seems weird to instantiate a class with a string when we want a fraction, and an int when we want an integer

File: test/arithmetic_test.rb [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+    assert_equal Magic.new(2, :magic_missile), @four / @two
+  end
+
+  test "#/ should multiply a number to the value" do
```
Comment (robfoster) [inline test/arithmetic_test.rb]: s/multiply/divide/s everywhere

File: test/arithmetic_test.rb (hunk 14 above)
Comment (qq99) [inline test/arithmetic_test.rb]: Disregard the above; I do think this should change the unit since you support multiplying unit by integer below (which you should use if you want to convey "twice as long")

File: test/arithmetic_test.rb (hunk 13 above)
Comment (kmcphillips) [inline test/arithmetic_test.rb]: Yes you can.

```
pry(main)> Measured::Weight.new(1, :lb) - Measured::Weight.new(1, :g)
=> #<Measured::Weight: 0.99779537738E0 lb>
```

What this is saying is that you can't subtract an inch from a pound.

File: lib/measured/units/dimension.rb (hunk 12 above)
Comment (kmcphillips) [inline lib/measured/units/dimension.rb]: From what I have been able to find, this is an exact conversion.

File: lib/measured/units/weight.rb (hunk 10 above)
Comment (kmcphillips) [inline lib/measured/units/weight.rb]: :+1:

File: lib/measured/units/dimension.rb (hunk 9 above)
Comment (kmcphillips) [inline lib/measured/units/dimension.rb]: Easy to do.

File: lib/measured/unit.rb (hunk 7 above)
Comment (kmcphillips) [inline lib/measured/unit.rb]: Forgot the message. Thanks.

File: test/arithmetic_test.rb (hunk 14 above)
Comment (kmcphillips) [inline test/arithmetic_test.rb]: It's not super useful. But I don't want to raise.

Mostly I want to support `Measured::Weight.new(1, :g) * 2`

File: test/arithmetic_test.rb (hunk 16 above)
Comment (kmcphillips) [inline test/arithmetic_test.rb]: This is simply the `BigDecimal` API. 

http://ruby-doc.org//stdlib-2.1.0/libdoc/bigdecimal/rdoc/BigDecimal.html

**Comment by robfoster** [comment]
WARN: Unresolved specs during Gem::Specification.reset:
      json (>= 1.7.7, ~> 1.7)
WARN: Clearing out unresolved specs.

File: test/conversion_test.rb [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+    @conversion.set_base :m, aliases: [:metre]
+
+    assert_raises Measured::UnitError do
+      @conversion.set_base :in
+    end
+  end
+
+  test "#add adds a new unit" do
+    @conversion.set_base :m
+    @conversion.add :in, aliases: [:inch], value: "0.0254 meter"
```
Comment (robfoster) [inline test/conversion_test.rb]: would be slick to be able to omit the aliases array when there's a single imput:
`@conversion.add :in, aliases: :inch, value: "0.0254 meter"`

File: test/arithmetic_test.rb (hunk 14 above)
Comment (qq99) [inline test/arithmetic_test.rb]: Should we remove `unit * unit` and only allow `unit * (int|float|decimal)` ?

File: lib/measured/conversion.rb [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+  def unit_names
+    @units.map{|u| u.name}.sort
+  end
+
+  def unit_or_alias?(name)
+    unit_names_with_aliases.include?(name.to_s)
+  end
+
+  def unit?(name)
+    unit_names.include?(name.to_s)
```
Comment (robfoster) [inline lib/measured/conversion.rb]: `Measured::UnitError: Cannot find unit for KG.`

Default to case sensitive unit name matches might be restrictive for no reason.

File: test/units/weight_test.rb [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+  test ".convert_to from g to kg" do
+    assert_conversion Measured::Weight, "2000 g", "2 kg"
+  end
+
+  test ".convert_to from g to lb" do
+    assert_conversion Measured::Weight, "2000 g", "0.4409245244E1 lb"
+  end
+
+  test ".convert_to from g to oz" do
+    assert_conversion Measured::Weight, "2000 g", "70.547923904 oz"
```
Comment (robfoster) [inline test/units/weight_test.rb]: Quantity[2000.0000001, "Grams"] http://www.wolframalpha.com/input/?i=70.547923904+oz+in+g
Hmm. I wonder if it's because of the 1 oz = 0.45359237⁄16 kg rational

`In the international yard and pound agreement of 1959 the United States and countries of the Commonwealth of Nations agreed to define the international avoirdupois ounce to be exactly 0.45359237⁄16 kg (28.349523125 g) by definition.` http://en.wikipedia.org/wiki/Ounce

File: lib/measured/units/dimension.rb [hunk 21]
```diff
@@ -0,0 +1,26 @@
+class Measured::Dimension < Measured::Measurable
```
Comment (robfoster) [inline lib/measured/units/dimension.rb]: NIST Special Publication 1038: The International System of Units (SI) – Conversion Factors for General Use (http://www.nist.gov/pml/wmd/metric/upload/SP1038.pdf ) is a document I was double checking our conversion with. Looks good!

But they refer to dimension as length, for example, "the kilometer measures length". .

**Comment by robfoster** [comment]
Added a few comment I hope were interesting but nothing blocking. :shipit: Nice work.

File: lib/measured/units/dimension.rb (hunk 21 above)
Comment (qq99) [inline lib/measured/units/dimension.rb]: I would think it would be `Distance` ?

Comment (qq99) [inline lib/measured/units/dimension.rb]: `dimension of a mathematical space (or object) is informally defined as the minimum number of coordinates needed to specify any point within it` https://en.wikipedia.org/wiki/Dimension

File: test/units/weight_test.rb (hunk 20 above)
Comment (qq99) [inline test/units/weight_test.rb]: Is this a problem?  I think sigdig wise, this is alright

**Comment by kmcphillips** [comment]
Fixed a bunch of comments. Thanks for all the feedback.

I'm going to go forward with how this is structured, despite some rounding anomalies and precision loss. In a subsequent PR I'm going to circle around some ways to reduce precision loss. One of my best ideas is to see if I can leave the value always represented as a `Rational` if any of the composite factors are rational. Then let the consumer of the unit do their own coercion, rounding, and precision loss.

Alternately I can change the tests to assert a certain level of rounding and just deal with the fact that we can never be completely perfect. We are already keeping more digits of precision than are specified in the standards.

File: lib/measured/conversion.rb (hunk 19 above)
Comment (kmcphillips) [inline lib/measured/conversion.rb]: Oh interesting. May be worth adding later.

**Comment by qq99** [comment]
:+1:

File: lib/measured/unit.rb [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+        other.conversion_amount <=> @conversion_amount
+      end
+    else
+      @name <=> other
+    end
+  end
+
+  def inverse_conversion_amount
+    if conversion_amount.is_a?(Rational)
+      Rational(conversion_amount.denominator, conversion_amount.numerator)
```
Comment (kmcphillips) [inline lib/measured/unit.rb]: I did make this change to prevent `Rational` from getting prematurely casted to a `BigDecimal`, though in the calls to multiply the value by the factor it is always eventually returned as `BigDecimal`. But this is slightly better.

File: lib/measured/units/dimension.rb (hunk 11 above)
Comment (kmcphillips) [inline lib/measured/units/dimension.rb]: Since this is parsed as `BigDecimal("0.01")` it is just as precise. Nothing is this lib is ever treated as a floating point number.

File: lib/measured/units/dimension.rb (hunk 21 above)
Comment (kmcphillips) [inline lib/measured/units/dimension.rb]: Missed this comment. Let me think about this. Good feedback.

**Comment by kmcphillips** [comment]
Renamed `Dimension` to `Length` to match standards:

screen shot 2015-03-05 at 10 01 31 am

**Comment by kmcphillips** [comment]
Ok. I really appreciate the feedback. I'm going to merge this today, release the gem, and then it can be iterated on. I've opened some issues to track that. #2 #3 #4

**Comment by maartenvg** [comment]
Haven't reviewed in great detail (didn't check each and every operation), but overall it looks complete and elegant. :sparkles: for owning this problem area.

:shipit:! While working with it we'll find any edge cases we've missed.

**Comment by kmcphillips** [comment]
Appreciate it!

**Comment by qq99** [comment]
:tada:

**Comment by garethson** [comment]
:clap:!
